DESIGN.md

```diff
@@ -28,6 +28,196 @@ daemon is spawned to handle future requests in its stead.
 
+## Versioning
+
+This section describes S3 Server's support for the specific AWS S3 Bucket
+Versioning feature.
+
+### AWS S3 Bucket Versioning
+
+An object in a versioning enabled bucket can have multiple versions, whereas an
+object can have only a single version in a standard bucket without versioning.
+
+With the standard S3 API, any update (PUT, DELETE) to an object in a versioning
+enabled bucket would create a new version of that object. Read operations such
+as GET and LIST would see the latest version of the objects. The table below
+shows such example, a `PUT foo` generates a new version `v2` and keeps `v1`;
+GET and LIST return the latest version of `foo`, which is `v2` in this case.
+
+| key | version | value |
+|-----|:-------:|:-----:|
+| foo |   v2    |   A   |
+| foo |   v1    |   B   |
+
+To work with all existing versions of the objects, S3 provides an extended
+API which uses the same operations of the standard API, but with versioning
+information specified. For example, in order to get the version `v1` of `foo`,
+users have to specify that version in the query, such as
+`GET /foo?versionId=v1`.
+
+A bucket can be non-versioning or versioning. The version of the objects of a
+bucket before it was versioning-enabled are called the "null" version; these
+versions are kept when a bucket is versioning enabled and are identified with
+the version identifier as "null". Suspending the versioning of a bucket would
+result in the overwrite of any previous null version when a new version is put,
+but would not delete all existing versions.
+
+### Implementation of Bucket Versioning in S3
+
+Each version of an object is stored as a separate key in metadata that
+corresponds to the data of the version in the datastore. The API component of
+S3 manages the put and delete of version data, sends instructions to create,
+retrieve, overwrite or delete a version's metadata to the metadata backend, and
+tracks extra information about a version, such as whether it is a delete marker
+or a null version.
+
+The metadata backend is responsible for handling the create, overwrite and
+delete operations of version metadata upon receiving instructions. We use
+version identifiers as the suffix for the keys of the object versions, and a
+special version to represent the latest version.
+
+The following is a summary of the design and details of implementation of
+versioning in S3.
+
+#### Version Id Format
+
+The version id is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3. The following information is encoded in a
+version ID:
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the texture format as metadata orders keys alphabetically
```

Comment:
It's the initial wording from the original document, I inferred it refers to the specific format for a version ID. I excluded the specific format for a version ID since I didn't sense the need for that level of detail, but since we're now including an example of the unencoded version IDs in the previous section, I'm adding it back.